include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+   * @param mpi_communicator MPI communicator over which logging operations are
+   * computer.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data,
+                          const MPI_Comm &      mpi_communicator);
+
+  /**
+   * Select a new nonlinear solver. All solver names used in this class are
+   * all lower case.
+   */
+  void
+  select(const typename AdditionalData::SolverType &type);
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+  void
+  set_data(
+    const typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+      &                                         additional_data,
+    const Teuchos::RCP<Teuchos::ParameterList> &parameters =
+      Teuchos::rcp(new Teuchos::ParameterList));
+#endif
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_SUNDIALS
+  void
+  set_data(const typename SUNDIALS::KINSOL<VectorType>::AdditionalData
+             &additional_data);
+#endif
+
+#ifdef DEAL_II_WITH_PETSC
+  void
+  set_data(const typename PETScWrappers::NonlinearSolverData &additional_data);
+#endif
+
+  /**
+   * Solve the nonlinear system. KINSOL uses the content of
+   * `initial_guess_and_solution` as an initial guess, and
+   * stores the final solution in the same vector.
+   *
+   * The functions herein are nearly identical in setup to what can be found
+   * in the KINSOL and NOX documentation.
+   */
+  void
+  solve(VectorType &initial_guess_and_solution);
+
+  /**
+   * A function object that users need to supply and that is intended to
+   * reinitize the given vector to its correct size, and block structure (if
+   * block vectors are used), along with any
+   * other properties necessary.
+   */
+  std::function<void(VectorType &)> reinit_vector;
+
+  /**
+   * A function object that users should supply and that is intended to
+   * compute the residual `dst = F(src)`.
+   *
+   * This function should return an int for either failure or success.
+   */
+  std::function<int(const VectorType &src, VectorType &dst)> residual;
+
+  /**
+   * A function object that users may supply and that is intended to
+   * prepare the linear solver for subsequent calls to
+   * solve_jacobian_system().
+   *
+   * The job of setup_jacobian() is to prepare the linear solver for
+   * subsequent calls to solve_with_jacobian(), in the solution of linear
+   * systems $Ax = b$. The exact nature of this system depends on the
+   * SolutionStrategy that has been selected.
+   *
+   * In the cases strategy = SolutionStrategy::newton or
+   * SolutionStrategy::linesearch, $A$ is the Jacobian $J = \partial
+   * F/\partial u$. If strategy = SolutionStrategy::picard, $A$ is the
+   * approximate Jacobian matrix $L$.
+   *
+   * The setup_jacobian() function may call a user-supplied function, or a
+   * function within the linear solver module, to compute Jacobian-related
```

Comment:
what is the "linear solver module"? I would just remove the second half of the sentence.